### PR TITLE
Publish public methods

### PR DESCRIPTION
**Pull request by n1313**
This change adds (basic) support for documenting public methods of components using jsdoc tags `@public`, `@param` and `@returns`.

### CODE REVIEW

File: src/rsg-components/ReactComponent/ReactComponent.js [hunk 1]
```diff
@@ -10,12 +11,16 @@ export default function ReactComponent({
 }) {
 	const { name, pathLine, examples } = component;
 	const { description, props } = component.props;
+	const methods = component.props.methods ? component.props.methods.filter((method) => {
```
Comment (sapegin) [inline src/rsg-components/ReactComponent/ReactComponent.js]: Maybe it’d be better to filter in the loader.

File: src/rsg-components/Methods/Methods.css [hunk 2]
```diff
@@ -0,0 +1,36 @@
+.table {
```
Comment (sapegin) [inline src/rsg-components/Methods/Methods.css]: We need to extract table with most of this CSS to a separate component and use it for props and methods.

File: src/rsg-components/Methods/MethodsRenderer.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import s from './methods.css';
+
+function renderRows(methods) {
+	const rows = [];
+	methods.map((method) => {
+		rows.push(
+			<tr key={method.name}>
+				<td className={s.cell}><Code className={s.name}>{method.name}()</Code></td>
+				<td className={s.cell}>{renderParameters(method)}</td>
+				<td className={s.cell + ' ' + s.cellDesc}>
```
Comment (sapegin) [inline src/rsg-components/Methods/MethodsRenderer.js]: Better to use `classnames` instead of concatenation.

File: src/rsg-components/Methods/Methods.css (hunk 2 above)
Comment (n1313) [inline src/rsg-components/Methods/Methods.css]: I thought of that too, but I am not sure this should be done as a part of this PR.

Comment (sapegin) [inline src/rsg-components/Methods/Methods.css]: It introduces duplication, so that’d make sense. But it’s up to you ;-)

File: src/rsg-components/Methods/MethodsRenderer.js (hunk 3 above)
Comment (n1313) [inline src/rsg-components/Methods/MethodsRenderer.js]: I agree, but this is how it is done in PropsRenderer which I used as a base for my PR. I decided to not deviate from the "style" of it too much. I think a classname refactoring should be done as a separate PR across the entire codebase.

Comment (sapegin) [inline src/rsg-components/Methods/MethodsRenderer.js]: Yeah, don’t worry, it’s just a minor thing ;-)

File: src/rsg-components/ReactComponent/ReactComponent.js (hunk 1 above)
Comment (sapegin) [inline src/rsg-components/ReactComponent/ReactComponent.js]: We already have some methods processing in the loader and we never need private methods, so why not send only the data that’s needed by the frontend. But that’s just my opinion ;-)

**Comment by n1313** [comment]
I can try to move it to the `next` branch, but I'd really like to have this change in v4 as well. Is there a way to achieve this painlessly?

**Comment by sapegin** [comment]
Painlessly for me — send two PRs for both branches ;-) But why do you want it in 4.x? 5.x needs testing, more time we spend in 4.x, more we postpone 5.x final release. I already use it in my project without any issues but I’d like more people to try it.

**Comment by n1313** [comment]
I have fixed two out of three things you've commented on and fixed the broken build (very embarrassing). I would very much like to have this change in v4 because I haven't touched v5 at all and have no idea how it works, or what the migration process is, or how difficult it will be to merge this PR into `next`. I will take a look at it as soon as possible, but I can't promise anything b/c I have no clue about v5 codebase at this moment.

**Comment by sapegin** [comment]
Thanks, merged!

I’m going to write a changelog for 5.x in the next few days (it’s going to be hard — there are 144 commits already).

**Comment by n1313** [comment]
Thanks!

**Comment by sapegin** [comment]
Looks like you broke something:

image 2017-01-09 at 6 23 35 pm

**Comment by n1313** [comment]
That's because I'm a doofus and empty arrays are truthy.
